### PR TITLE
PEX-15769 fix fortify findings

### PR DESCRIPTION
**Pull request by jmarten**

### CODE REVIEW

**Review by hua-sig — APPROVED** [review]

File: data/src/main/java/smile/data/parser/IOUtils.java [hunk 1]
```diff
@@ -30,24 +30,28 @@
...
+        return file;
     }
-
+    
     /** Get the reader of sample datasets. */
     public static BufferedReader getTestDataReader(String path) throws FileNotFoundException {
         return new BufferedReader(new InputStreamReader(new FileInputStream(getTestDataFile(path))));
     }
-
+
```
Comment (hua-sig) [inline data/src/main/java/smile/data/parser/IOUtils.java:54]: trivial: some tabs/whitespaces were added

File: data/src/main/java/smile/data/parser/microarray/PCLParser.java [hunk 2]
```diff
@@ -127,7 +129,7 @@ public AttributeDataset parse(File file) throws IOException, ParseException {
      * @throws java.io.IOException
      */
     public AttributeDataset parse(String name, File file) throws IOException, ParseException {
-        return parse(name, new FileInputStream(file));
+        return parse(file);
```
Comment (jmarten) [inline data/src/main/java/smile/data/parser/microarray/PCLParser.java:132]: Do we discard the name here on purpose?

Comment (rohini2021) [reply data/src/main/java/smile/data/parser/microarray/PCLParser.java:132]: It was a change done by you to fix compilation errors I think. @jmarten

Comment (rohini2021) [reply data/src/main/java/smile/data/parser/microarray/PCLParser.java:132]: I can add it back if it was by accident

File: nlp/src/main/java/smile/nlp/dictionary/SimpleDictionary.java [hunk 3]
```diff
@@ -44,9 +45,11 @@ public SimpleDictionary(String resource) {
...
 
         File file = new File(resource);
-        try (BufferedReader input = file.exists() ?
-             new BufferedReader(new FileReader(resource)) :
-             new BufferedReader(new InputStreamReader(this.getClass().getResourceAsStream(resource)))) {
+        try (FileReader fileReader = file.exists() ? new FileReader(resource) : null;
+                InputStream resourceAsStream = this.getClass().getResourceAsStream(resource);
+                InputStreamReader inputStreamReader = new InputStreamReader(resourceAsStream);
+                BufferedReader input = file.exists() ? new BufferedReader(fileReader) :
+                        new BufferedReader(inputStreamReader)) {
```
Comment (hua-sig) [inline nlp/src/main/java/smile/nlp/dictionary/SimpleDictionary.java:52]: why do we change here? The original version is already using try with resources, right? I think the reader will close the underlying inputstream, is the fortify check complaining about the inputstream wrapped by the reader not being closed?

Besides the original version only creates inputStream-/Reader if the file does not exist, while the changed version will create it regardless.

Comment (rohini2021) [reply nlp/src/main/java/smile/nlp/dictionary/SimpleDictionary.java:52]: Yes, fortify check complains if we dont have separate resource declaration. Also, I think only the declared resources will be deleted

**Review by hua-sig — APPROVED** [review]

**Pull request merged**